framework/sysdeps/windows/tmpfile.c

```diff
@@ -40,9 +41,68 @@ static unsigned next_random()
     return result;
 }
 
+static bool is_dir_writable(const wchar_t *dir)
+{
+    if (!dir)
+        return false;
+
+    size_t len = wcslen(dir);
+    if (!len)
+        return false;
+    if (dir[len - 1] == L'\\')
+        return _waccess(dir, W_OK) == 0;
+
+    // append a backslash so we confirm it's a directory
+    wchar_t path[MAX_PATH];
+    wcscpy(path, dir);
+    wcscat(path, L"\\");
+    return _waccess(path, W_OK) == 0;
+}
+
+static const wchar_t *find_alternate_temp_dir()
+{
+    static const wchar_t *dirname = NULL;
+    if (dirname)
+        return dirname;
+
+    dirname = _wgetenv(L"TEMP");
+    if (is_dir_writable(dirname))
+        return dirname;
+
+    dirname = _wgetenv(L"TMP");
+    if (is_dir_writable(dirname))
+        return dirname;
+
+    return NULL;
+}
+
+// fills in the temporary directory in tmpname and returns the length of the path
+// (including the terminating backslash)
+static size_t fill_tmpdir(wchar_t tmpname[MAX_PATH])
+{
+    size_t len = GetTempPathW(MAX_PATH, tmpname);
+    if (len) {
+        wcscat(tmpname, L"\\");
```

Comment:
`GetTempPathW` always returns a string that ends with a backslash; this step is not really needed. Also, if string with length equal to `MAX_PATH` was returned, this step would overflow `tmpname`.

Reply:
Didn't notice that it returned a string ending in backslash in the documentation, but it's right there. Thanks, adapting.